### ISSUE TITLE
define a set of analysis options shared by analyzer packages

### ISSUE THREAD

**Issue by pq** [dart-lang/sdk#53165]
See the discussion in https://github.com/dart-lang/linter/pull/4653.

Assuming we can agree on a core set of shared lints, it would be great to codify them in a singled shared set of options.

/fyi @bwilkerson @scheglov @srawlins 

(Apologies if I've filed this issue before -- I can't find it...)

/fyi @devoncarew in case there's been any progress on a Dart ecosystem options set

**Comment by parlough** [comment]
Starting from https://github.com/dart-lang/ecosystem/tree/main/pkgs/dart_flutter_team_lints would be a good start, at least for consistency. It might require some ignores or larger migrations though. Any extra can be added on top of that.

**Comment by devoncarew** [comment]
I would definitely try to base the analyzer lint config off of package:dart_flutter_team_lints (and, it would be interesting to know what additional lints are turned on above and beyond what dart_flutter_team_lints defines). package:dart_flutter_team_lints is already dep'd into the sdk so you should be all set.

**Comment by pq** [comment]
/fyi @kenzieschmoll, @elliette  as ideally we could come up with a plan that includes alignment with `devtools` as well.

**Comment by kenzieschmoll** [comment]
DevTools uses `flutter_lints` and several lints from DCM currently: https://github.com/flutter/devtools/blob/master/packages/analysis_options.yaml. My initial thought is that DevTools will likely want to use lints that are more specific to writing a Flutter app that the analyzer packages may not care about, but that hypothesis would need to be validated by comparing what is in dart_flutter_team_lints to what is in flutter_lints + our DCM lints configuration.

**Comment by bwilkerson** [comment]
My hypothesis is that the closer we can get to having a single unified coding style, the easier it will be for members of our team to move between code bases.

Enabling flutter-specific lints won't cause any problems in code that isn't using Flutter, so I'm not sure that's an issue. I suspect that it would be a no-op to enable `flutter_lints` in all of our packages.

I'm not familiar with the DCM lints that are enabled (and the link above results in a 404 for me so I don't even have the names of the lints as a reference), but it does seem like we should discuss whether we want them enabled for non-DevTools code.

At the very least it does seem like we ought to be able to have a set of lints that we use across all of our code bases that's defined in a single location, even if there are additional lints for some packages that we can't agree to apply everywhere.

**Comment by incendial** [comment]
That file was moved here after the migration to workspaces : https://github.com/flutter/devtools/blob/master/analysis_options.yaml

> but it does seem like we should discuss whether we want them enabled for non-DevTools code.
Can you please elaborate on what "enabled" means in this context as adding the rules to "analysis_options" won't be sufficient for them to be actually enabled?

**Comment by bwilkerson** [comment]
> Can you please elaborate on what "enabled" means in this context as adding the rules to "analysis_options" won't be sufficient for them to be actually enabled?

I meant "adding a 'dart_code_metrics/rules' section to the analysis options files that lists the agreed on set of DCM lints".

The question is whether we want the `analyzer`, `analysis_server`, `dartdoc`, `linter` (and a few other packages) to use the same set of lints as those used by `flutter/devtools`. The DCM lints aren't currently enabled for any of those packages.

**Comment by incendial** [comment]
Thanks! Do you have any plans to enable them in those other packages (AFAIK, no)? If not, then just listing them should not affect any other package, right? Though if your plan is to not add them because they are used only by `flutter/devtools` (and have that repo list them while including your common config), then I understand the point.

Edit: Ah, and if someone with enabled DCM tries to contribute, they will get warnings and that will be very confusing. So I get your point even more.

I'm just trying to understand whether your words are actionable feedback for us or not (looks like not).

**Comment by bwilkerson** [comment]
No, I don't think there's anything we need from DCM, but thanks for being on top of it!

I think this issue is just a task for us to make some decisions about which lints are enabled across our combined code bases in light of the fact that our combined team should be expecting to work across package boundaries that we previously weren't working across.
